### ISSUE TITLE
Exodus purchases, DEx payments, crowdsales unconfirmed transactions

### ISSUE THREAD

**Issue by dexX7** [OmniLayer/omnicore#73]
Over the last days I tried to make up my mind about how BTC related actions can be nicely integrated into Omni Core, because right now BTC doesn't really fit into the model.

There are a few points I'd like to discuss.

1) Exodus purchases and DEx payments are BTC payments, but they are handled differently.

It's probably not too far fetched to consider Exodus purchases and DEx payments as something similar, at least design wise. There are some significant differences though:

 - A DEx payment requires a marker (Exodus output, or "om" bytes), while an Exodus purchase, strictly seen, doesn't. This is only relevant for testnet and regtest mode [extra question: why do use a different destination, which is not the marker here?], because on mainnet the Exodus marker is also the fundraiser destination. However, it's worth to notice that a BTC payment to `moneyqMan7uh8FqdCA2BV5yZ8qVrc9ikLP` can count as Exodus purchase, even without explicit marker.

 - A DEx payment is only considered as such, if the transaction does not embed any payload, i.e. the transaction has a marker, but isn't a class A, B or C transaction. In contrast, the Exodus purchase logic is executed in any case, even if there is another "action" bound to the transaction (such as a simple send, as seen on mainnet).

 - There can only be one Exodus purchase per transaction, namely the highest amount sent to Exodus is used to determine the purchased amount, while at the same time, there can be several DEx payments, and for each output on it's own the DEx payment logic is executed (whether it's more than one payment to the same seller, or multiple sellers).

I would really like, if we come up with an approach, which is more consistent, so that Exodus purchases and DEx payments might be handled by the same module. Given that we can't change history, we have to build a model around those limitations, or change the rules for the future.

2) Exodus purchases are crowsale purchases

 - That's a very interesting point in my opinon: what exactly is an Exodus purchase? Isn't it a crowdsale, with 10 % early bird bonus per week, with a block height based deadline at block 255365, and a time based deadline at 1377993600 (2013-09-01 00:00:00 UTC)? ;)

 - The unfortunate point here is that the deadline block 255365 and the deadline timestamp 1377993600 don't refer to the same point in time, and block 255365 actually has a timestamp of 2013-09-01 00:04:34 UTC.

This alone isn't a deal breaker, because we might introduce the concept (code wise) of an actual deadline (in terms of: end of crowdsale), and a deadline for the bonus calculations. For all tx 51 crowdsales those deadlines are time based, and equal.

3) Floating-point numbers and rounding directions ...

 - The Meta-DEx and STO transactions are completely unaffected by any floating point issues, and use plain integer math. Results are either rounded up, or down (i.e. `1.00001 -> 2 or 1`).

 - The tx 51 crowdsale logic uses integer math for intermediate calculations, but has a concept of fixed precision, and results are divided by `1000000000000L` and `100000000` to limit the number of digits. I'm not yet sure, what this means in practise, and whether this implies a loss of precision.

 - Unfortunally the DEx payment logic, as well as the Exodus purchase logic use `double` for the calculations, and results are "rounded half up" (i.e. `0.49 -> 0, 0.50 -> 1`).

I started to test the tx 51 crowdsale logic with Exodus fundraiser values, and at least for some input, the results differ by one satoshi. I have no idea, whether this applies to mainnet purchases, or if we may get away in practise with using the crowdsale logic to handle Exodus purchases.

It would be incredibly awesome, if we could get away with it, because then we could get rid of any special "Exodus purchase logic", and simply consider Exodus purchases as crowdsale purchases.

---

My intention is to find a model, where BTC payments fit in nicely, while thinking ahead, where we have to deal with BTC based crowdsales, and maybe a traditional DEx v2 (e.g. #253: Adaptive DEx fees based on purchased amount, #184: Creating MSC/BTC options via DEx fees to the seller).

I'm a bit lost, because I'm not sure, how to continue from here, but I think finding answers to the following questions might help:

1. Do we want to continue to consider DEx payments only as such, if the transaction doesn't embed a payload?

2. How are BTC based crowdsales going to be handled? Might it be possible to participate in several BTC based crowdsales in one transaction, similar to DEx payments, where each output may be a participation/purchase/payment?

3. Would it be possible to get rid of floating point numbers for DEx payments and Exodus purchases?

Code wise, I'd like to turn `parseTransaction()` into a plain "extract the payload, determine sender and reference" function, and execute any actions (DEx payments, Exodus purchases) somewhere else.

In light of the above, we currently have to determine:

 - (a) does the transaction have a marker? (relevant for class A, B, C, DEx payments - not relevant for Exodus purchases)
 - (a) which encoding class is used? (relevant for class A, B, C)
 - (b) who is the sender? (relevant for for class A, B, C, DEx payments, Exodus purchases)
 - (b) who is the reference? (relevant for for class A, B, C)
 - (b) does the transaction have a payload? (relevant for DEx payments)
 - (b) if so, what is the payload? (relevant for for class A, B, C)
 - (b) what is the transaction fee? (relevant for DEx accepts)
 - (c) does the transaction have a money output [1EXoDus, moneyqMan]? (relevant for Exodus purchases)
 - (c) what is the highest value sent to the money output? (relevant for Exodus purchases)

And we have to handle:

 - Exodus purchases (requires b, c)
 - DEx payments (requires a, b, c)
 - BTC crowdsales at some point (requires a, b, c)
 - standard payload-transactions (requires a, b, c)

As far as I can see we may split `parseTransaction()` and tackle (a), (b), (c) seperately:

 - (a) Determine, whether the transaction has a marker, or better: determine the encoding class (cheap)
 - (b) Extract payload, determine sender, reference, and transaction fees (expensive)
 - (c) Get highest money output [...] (cheap)

Iterating over outputs is smoking fast and "cheap", while retrieving inputs isn't, thus "expensive". There is probably no performance penalty, even if we check the existence of a marker twice, or collect output values for class A, and do another loop for Exdus purchases.

I further imagine we may persist the data gained by (b), in particular: `txid`, `tx fee`, `encoding class`, `sender`, `payload`, `reference` (is that all?), so we don't have to do the expensive parsing more than once. This is something that is going to be very handy for unconfirmed transactions, or a no-txindex mode as well. It raises the following additional questions:

4. If we are going to parse unconfirmed transactions, then we can only estimate in which block a transaction may end up. This isn't a huge problem, and we may persist the data we have, and update the entry (i.e. add block height, transaction position within block), once the transaction confirms. However, the problem I'm not sure about: the parsing logic explicitly uses a block height to determine, whether P2SH outputs are allowed, if class C transactions are enabled. We may assume that an unconfirmed transaction ends up in a block, which has a height greater than the current chain height, but it's still fuzzy during switch-over times. Say for example: current chain height is 356003, and class C transactions are activated at height 356005. Should an unconfirmed transaction be processed as class C transaction or not? Or maybe both cases might be handled, i.e. as class B and as class C? If so, where is the cutoff? 3 blocks before the switch-over, 5 blocks, 10 blocks, ...?

5. If `parseTransaction` processes transactions only once, and the data is persisted, then it's reasonable to properly persist information about Exodus purchases, and DEx payments, as well. This is already done to some degree, but not completely. Ideally this should be usable for BTC crowdsales as well. Persisting *any* output of transactions with marker/money output surely works, but it's an absurd overkill. So my last question: what is the minimum data do we need to persist, to cover that as well? I think a small overhead is fine, i.e. it's acceptable, if we store data, which *could* be related to Exodus purchases or DEx payments (even if it turns out it's not), but we probably don't have to persist data, which certainly won't be used for Exodus purchases [...], say for example, because the Exodus period is already over.

Ideally, I'd like to be able to start with a clean state, and then iterate over the persisted data, and end up with a similar state, as if the data wasn't retrieved from a DB, but extracted from transactions, as currently done.

Sorry in advance for the very long post, but especially finding answers to the last two questions would be very, very helpful for me.

**Comment by zathras-crypto** [comment]
Few comments from first read (I'm a bit fuzzy, will have another read later :))

* I consider Exodus purchases as a crowdsale.  I think it would be silly not to given the usage of crowdsale-esque parameters (early bird bonus, deadline etc etc)
* The issue with the deadline is that JR "allowed" a few late-comers to be included, which does complicate matters a little.  I believe this can be handled as a 'special case' to crowdsale logic, rather than completely separate Exodus purchase code as per current.
* We will be allowing BTC crowdsales at some point in future, and I believe Exodus purchases should be a derivation of this (some custom logic will be required, because two properties instead of one were issued and the deadline issue as above, but this should be minimal).
* I'd like to avoid FP math everywhere.
* Allowing DEx payments to also carry a payload and perform a secondary action seems overkill to me at this stage since we are not very well setup to handle multiple actions in the same tx (just look at simple sends + crowdsale participation, it's kind of hacky) - do you have an example use case?
* DEx payments to multiple parties in one transaction I dislike - but I had to build support for it - I can't actually remember why hehe.

I love where you're going with this, and I'm on board - but there are some caveats.

* We discussed 0.0.10 release this morning, it's likely going to be first half of July, with a 4 week period following that for integrators/users to update before 0.0.10 features go live and old versions fall out of consensus.  Thus our primary focus should be to make 0.0.10 as solid as it can be before that point.
* Unconfirmed transactions will never be 'usable'.  We can provide decoding and notification of unconfirmed transactions for sure, but we can never address the question of validity.  With Omni being a state based system it is not possible to know validity before the transaction is actually mined - the best we can do is say "so far we haven't *yet* seen anything in the mempool that would invalidate this transaction, but until it's mined we can't know for sure".  
* I'm a big fan of the idea of ditching txindex, storing our data outside the blockchain and allowing pruning - it will be a big boon to usability (requiring 40 odd gigs can only hurt user uptake) - we've just got to make sure to do it in a safe and secure way.

TL:DR; some very worthwhile points in your post dude, and a lot of them I'm all for.  I suggest once 0.0.10 is released and settled in we 'take stock' as it were and plan for 0.0.11 and what our priorities are going to be - a major re-architecture can be on the cards for sure if we can get the support to do it :)

**Comment by dexX7** [comment]
> Allowing DEx payments to also carry a payload and perform a secondary action seems overkill to me at this stage ...

> DEx payments to multiple parties in one transaction I dislike ...

Yeah, I'm just trying to find a common basis, so to speak. Supporting multiple DEx purchases, to multiple receivers, in one transaction is what's currently done. And Exodus purchases are not constrained by the fact, that a transaction may also have another action (there are a few mainnet transactions of simple sends, which are Exodus purchases at the same time).

Logic wise it would probably be easier to handle the BTC stuff completely seperated, ideally on an output-for-output basis (after passing some checks such as the marker check). Easier in terms of "not having to go through the whole parsing steps, only to see that a transaction is not a class A transaction, while making sure, it's also not just a RPC call".

Let's ignore for a moment everything that's currently done, and quickly think about a logic flow, which allows to address payload-transactions, Exodus purchases, DEx payments, etc. How would that look like? I'm not really sure.

> We discussed 0.0.10 release this morning, it's likely going to be first half of July, with a 4 week period following that for integrators/users to update

Awesome, that's great to hear!

Just some note to clarify: a feature freeze seems reasonable (unless it's low hanging), and it's not really my intention to include BTC logic and unconfirmed transactions in 0.0.10, but to get a better feeling for where it's going, which may have impliciations to the ongoing work.

Speaking of, something very specific we may discuss quickly:

 - As far as I can see, a transaction without marker, but only one or more money outputs passes the marker check ([omnicore.cpp#L765](https://github.com/OmniLayer/omnicore/blob/3d647f5492f9b3d464e8a5dc17b9a92df28303ac/src/omnicore/omnicore.cpp#L765)). I'm wondering: are testnet transactions without marker, but money output as marker replacement, considered as valid right now?

 - During the marker checks there are no explicit checks, whether some output type is allowed. What might happen, if there is a class B/multisig transaction on mainnet with "om" `OP_RETURN` marker, and without Exodus output (before `OP_RETURN` is enabled)? On the first glimpse it looks like it would pass the marker check (because it has "om" bytes), and would then be classified as class B transaction, because `OP_RETURN` isn't allowed yet.

**Comment by zathras-crypto** [comment]
> What might happen, if there is a class B/multisig transaction on mainnet with "om" OP_RETURN marker

That's a very good point actually - raises the question "should the marker be class specific?"

> but only one or more money outputs passes the marker

Yeah the moneyman thing is an interesting case because it's a deviation from what we do on mainnet to facilitate testing.  I'd probably prefer to scrap it altogether and have a Test MSC faucet, but that's probably just me hehe

**Comment by dexX7** [comment]
> "should the marker be class specific?"

In the context of class C it must, at least until the switch-over, to prevent that consensus is broken for old clients, if someone sends a transaction with class C marker, and class B payload.

> Yeah the moneyman thing is an interesting case ...

Do you know why this approach was chosen?

**Comment by zathras-crypto** [comment]
> In the context of class C it must, at least until the switch-over, to prevent that consensus is broken for old clients, if someone sends a transaction with class C marker, and class B payload.

Great point, in which case I suggest we make the marker class specific full stop.

> Do you know why this approach was chosen?

Probably to get rid of the old ```mastercoin_balances.txt``` file you used to have to populate to get a test MSC balance.  Since it is a deviation from what we do on mainnet however I think it's not ideal (I get why Michael did it this way though, I just think it's time to move to a less invasive approach)

**Comment by dexX7** [comment]
> Great point, in which case I suggest we make the marker class specific full stop.

Yeah, I agree. This is something I haven't tested though, but I assume, based on the code, the scenario mentioned above is possible, so we definetly should make the marker check `isAllowedOutputType`-aware.

> I just think it's time to move to a less invasive approach

Are you suggesting to drop the money approach, and use Exodus instead? I'd be in favor of it, but it would basically reset our testnet.

**Comment by zathras-crypto** [comment]
> so we definetly should make the marker check isAllowedOutputType-aware.

This would still permit a Class B transaction with a Class C marker once Class C has gone live though right?  I fear this introduces complexity we don't need...

> but I assume, based on the code

We'd need to test, but I would imagine it would go something like this:

* Existence of Class C marker passes the marker check
* Existence of OP_RETURN data output forces Class selection for parsing to Class C.   Class B multisig outputs ignored.
* Class C not enabled yet so parser throws out the transaction,

> Are you suggesting to drop the money approach, and use Exodus instead?

I haven't given much thought, just in general terms the moneyman approach means we approach testnet differently to mainnet.  As for Exodus, again not sure how we'd work that (perpetual crowdsale is again a deviation from mainnet).  A faucet that hands out TestNet BTC & TestNet MSC in one go would be a nice/elegant solution to this, but would require setting something up.

> I'd be in favor of it, but it would basically reset our testnet.

Well we're about to reset all the trades on testnet anyway so I'm not really fussed about that - but I don't have a thought out strategy for replacing sends to moneyman, I'm just commenting that having separate things for testnet can confuse things and ideally we'd not do that :)

**Comment by dexX7** [comment]
> This would still permit a Class B transaction with a Class C marker once Class C has gone live though right? I fear this introduces complexity we don't need...

> Existence of Class C marker passes the marker check [...]

Well, let's combine the marker check and class selection then?

> As for Exodus, again not sure how we'd work that (perpetual crowdsale is again a deviation from mainnet).

I don't think a never-ending Exodus crowdsale is a bad thing, and without one we'd basically have a testnet without Exodus sale, which also differs significantly from mainnet.

A faucet doesn't work for regtest mode.

**Comment by zathras-crypto** [comment]
> Well, let's combine the marker check and class selection then?

That sounds sensible.

> I don't think a never-ending Exodus crowdsale is a bad thing, and without one we'd basically have a testnet without Exodus sale, which also differs significantly from mainnet.
A faucet doesn't work for regtest mode.

Fair points :)  How would you suggest to handle things like bonus and deadline (or just ignore them as per moneyman and issue 100 TMSC per TBTC?)

**Comment by dexX7** [comment]
I wouldn't ignore them, but set a very high value for the deadline, an early bird bonus of 0, and a bonus of 0 for the issuer. Using a payout factor of 100 seems fine. In the end, this is probably the same, but handling the Exodus payouts as close to a crowdsale as possible would be great.

Since MSC and TMSC are emitted, maybe we should think of it as two crowdsales at the same time, which are both associated with the same address, and both take BTC etc.

**Comment by dexX7** [comment]
While looking at parseTransaction, the Exodus logic is as follows:

> Take the value of the *first* Exodus output

This seems to have no mainnet impact, as there were probably only transactions with one Exodus output, but we may take a closer look at this, given that there might be as well more than one Exodus output on testnet/regtest mode.

I'm not necessarily in favor of counting more than one Exodus purchase, if there are more than one Exodus outputs, but I don't really see why we should stick to the "take first" approach. What do you think?

In case of doubt, I tend to count each output on it's own (thus multiple purchases might be possible).

**Comment by dexX7** [comment]
I started with [wip-oc-0.10-parse-transaction-classes](https://github.com/dexX7/bitcoin/compare/7dde7dc71f247a1f95160d60d2bcce193985cad7...dexX7:wip-oc-0.10-parse-transaction-classes) to tackle the marker/class issue, and I'm probably going to move some parts here and there.

But a two questions:

 - How are we going to handle the Exodus/Money address on testnet? I tend to use only one, but this comes at the cost of resetting testnet.

 - How are multiple Exodus purchases in one transaction handled? I tend to allow more than one, and consider each output as "atomic action".

**Comment by zathras-crypto** [comment]
> How are we going to handle the Exodus/Money address on testnet? I tend to use only one, but this comes at the cost of resetting testnet.

> How are multiple Exodus purchases in one transaction handled? I tend to allow more than one, and consider each output as "atomic action".

I assume the benefit here is to avoid the extra moneyman code, so wouldn't it be prudent to use the same Exodus code as mainnet?  This can be changed to allow multiple Exodus outputs without any consensus issues.  As to whether to count each as an atomic action - again I don't feel we're particularly well equipped to handle multiple actions per single transaction in our current codebase.   Since these multiple actions are not different actions but the same action, why not just sum multiple outputs to Exodus and process as a single action?

As for the reset testnet thing - I'm not hugely concerned by that but we should probably ask if anyone else has any objections :)

**Comment by dexX7** [comment]
Regarding the MoneyMan code, I think we could also have the concept of a MoneyAddress, seperately to an ExodusAddress. On mainnet, the ExodusAddress and the MoneyAddress are the same, but on testnet they are not.

Assuming we maintain compability as far as possible (i.e. no testnet reset), then it raises again the question: is it allowed to have a Money output, but no Exodus marker? I think it shouldn't be.

> As to whether to count each as an atomic action - again I don't feel we're particularly well equipped to handle multiple actions per single transaction in our current codebase. Since these multiple actions are not different actions but the same action, why not just sum multiple outputs to Exodus and process as a single action?

Code wise, I'd like to get somehow close to something like:

```c++
for (unsigned int n = 0; n < tx.vout.size(); ++n) {
    CTxDestination dest;
    if (ExtractDestination(tx.vout[n].scriptPubKey, dest)) {
        if (CBitcoinAddress(dest) == moneyAddress) {
            TXExodusFundraiser(tx, strSender, tx.vout[n].nValue, nBlock, nTime);
            // if we only "take the first", then we may uncomment the break
            // break;
        }
    }
}
```

Currently Exodus purchases are pretty much unnoticed, and they are at best logged, but don't show up in the UI or on the RPC layer, so I'm not really concerned about "multi actions" at the moment. As said though, I'm also not necessarily in favor of allowing multiple purchases, but it's a bit similar to DEx purchases, and if we allow more than one per transaction, then I think we should also allow more than one Exodus purchase.

**Comment by zathras-crypto** [comment]
> Assuming we maintain compability as far as possible (i.e. no testnet reset), then it raises again the question: is it allowed to have a Money output, but no Exodus marker? I think it shouldn't be.

**Comment by dexX7** [comment]
> I don't really get that mate - some (I'd guess a majority) of existing sends to moneyman will not have an Exodus marker 

Ahh, sorry, this changes everything. All my previous posts were based on the assumption those purchases actually have an Exodus marker *and* Money output.

> Scenario B is simpler in my book and I don't really see a huge value to using

I guess my only argument against it is how DEx payments are handled, otherwise I agree.

**Comment by zathras-crypto** [comment]
> Ahh, sorry, this changes everything. All my previous posts were based on the assumption those purchases actually have an Exodus marker and Money output.

I'd need to check to be sure as I'm only going on memory here :)

> I guess my only argument against it is how DEx payments are handled, otherwise I agree.

Yeah I do see your point - I just remember when I was doing the DEx payment stuff that it felt ugly and hacky - I'm sure I didn't *want* to support payments to multiple sellers in the same transaction but I'm sure I had to for a reason (perhaps consensus with msc-tools or something?).

Longer term I absolutely agree that we should support multiple actions per transaction (example send-to-many) but I think this should come as a kind of 'rearchitecture' that we should queue up next - ie get 0.0.10 out and then press 'pause' and take stock of our current position and limitations, and see about a rearchitecture to address.  This is where things like no-txindex/pruning, storing our transactions outside the blockchain and so on come in, and we can redesign with a year of experience in mind rather than the current additive model.

**Comment by dexX7** [comment]
Good points, you're right. I shouldn't necessarily start something new now, so I'm just going to stick to the current behavior, and address the class/marker check issue. And maybe do some non-functional changes on the way. :)
